test/units/testsuite-02.sh

```diff
@@ -34,77 +34,50 @@ fi
 # Arguments:
 #   $1: test path
 #   $2: test exit code
-report_result() {
-    if [[ $# -ne 2 ]]; then
-        echo >&2 "check_result: missing arguments"
+run_test() {
+    if [[ $# -ne 1 ]]; then
+        echo >&2 "run_test: missing arguments"
         exit 1
     fi
 
-    local name="${1##*/}"
-    local ret=$2
+    local test="$1"
+    local name="${test##*/}"
+
+    echo "Executing test $name as unit $name.service"
+
+    systemd-run --quiet --property Delegate=1 --unit="$name" --wait "$test" && ret=0 || ret=$?
+
+    exec 420</lock
+    flock 420
 
     if [[ $ret -ne 0 && $ret != 77 && $ret != 127 ]]; then
         echo "$name failed with $ret"
         echo "$name" >>/failed-tests
         {
             echo "--- $name begin ---"
-            cat "/$name.log"
+            journalctl --unit="$name" --no-hostname -o short-monotonic
             echo "--- $name end ---"
         } >>/failed
     elif [[ $ret == 77 || $ret == 127 ]]; then
         echo "$name skipped"
         echo "$name" >>/skipped-tests
         {
             echo "--- $name begin ---"
-            cat "/$name.log"
+            journalctl --unit="$name" --no-hostname -o short-monotonic
             echo "--- $name end ---"
         } >>/skipped
     else
         echo "$name OK"
         echo "$name" >>/testok
     fi
-}
 
-set +x
-# Associative array for running tasks, where running[test-path]=PID
-declare -A running=()
-for task in "${TEST_LIST[@]}"; do
-    # If there's MAX_QUEUE_SIZE running tasks, keep checking the running queue
-    # until one of the tasks finishes, so we can replace it.
-    while [[ ${#running[@]} -ge $MAX_QUEUE_SIZE ]]; do
-        for key in "${!running[@]}"; do
-            if ! kill -0 "${running[$key]}" &>/dev/null; then
-                # Task has finished, report its result and drop it from the queue
-                wait "${running[$key]}" && ec=0 || ec=$?
-                report_result "$key" "$ec"
-                unset "running[$key]"
-                # Break from inner for loop and outer while loop to skip
-                # the sleep below when we find a free slot in the queue
-                break 2
-            fi
-        done
-
-        # Precisely* calculated constant to keep the spinlock from burning the CPU(s)
-        sleep 0.01
-    done
-
-    if [[ -x $task ]]; then
-        echo "Executing test '$task'"
-        log_file="/${task##*/}.log"
-        $task &>"$log_file" &
-        running[$task]=$!
-    fi
-done
+    exec 420<&-
```

Comment:
The corresponding fd close.

```suggestion
    exec "$LOCK_FD"<&-
```